rust/agama-network/src/dbus/tree_manager.rs

```diff
@@ -0,0 +1,164 @@
+        let mut objects = self.objects.lock();
+
+        let path = format!(
+            "/org/opensuse/Agama/Network1/connections/{}",
+            objects.connections.len()
+        );
+        let cloned = Arc::new(Mutex::new(conn.clone()));
+        self.add_interface(
+            &path,
+            interfaces::Connection::new(Arc::clone(&self.network), Arc::clone(&cloned)),
+        )
+        .await?;
+
+        self.add_interface(
+            &path,
+            interfaces::Ipv4::new(Arc::clone(&self.network), Arc::clone(&cloned)),
+        )
+        .await?;
+
+        if let Connection::Wireless(_) = conn {
+            self.add_interface(
+                &path,
+                interfaces::Wireless::new(Arc::clone(&self.network), Arc::clone(&cloned)),
+            )
+            .await?;
+        }
+
+        objects.register_connection(conn.uuid(), &path);
+        Ok(())
+    }
+
+    /// Removes a connection from the tree
+    pub async fn remove_connection(&mut self, uuid: Uuid) -> Result<(), ServiceError> {
+        let mut objects = self.objects.lock();
+        let path = objects.connection_path(uuid).unwrap();
+        let object_server = self.connection.object_server();
+        _ = object_server.remove::<interfaces::Wireless, _>(path).await;
+        object_server.remove::<interfaces::Ipv4, _>(path).await?;
+        object_server
+            .remove::<interfaces::Connection, _>(path)
+            .await?;
+        objects.unregister_connection(uuid).unwrap();
+        Ok(())
+    }
+
+    async fn add_interface<T>(&self, path: &str, iface: T) -> Result<bool, ServiceError>
+    where
+        T: zbus::Interface,
+    {
+        let object_server = self.connection.object_server();
+        Ok(object_server.at(path.clone(), iface).await?)
+    }
+}
+
+/// Objects paths for known devices and connections
+#[derive(Debug, Default)]
+pub struct ObjectsRegistry {
+    pub devices: HashMap<String, String>,
+    pub connections: HashMap<Uuid, String>,
+}
+
+impl ObjectsRegistry {
+    pub fn register_device(&mut self, name: &str, path: &str) {
+        self.devices.insert(name.to_string(), path.to_string());
+    }
+
+    pub fn register_connection(&mut self, uuid: Uuid, path: &str) {
+        self.connections.insert(uuid, path.to_string());
+    }
+
+    pub fn device_path(&self, name: &str) -> Option<&str> {
+        self.devices.get(name).map(|p| p.as_str())
+    }
+
+    pub fn connection_path(&self, uuid: Uuid) -> Option<&str> {
+        self.connections.get(&uuid).map(|p| p.as_str())
+    }
+
+    pub fn unregister_device(&mut self, name: &str) -> Option<String> {
+        self.devices.remove(name)
+    }
+
+    pub fn unregister_connection(&mut self, uuid: Uuid) -> Option<String> {
+        self.connections.remove(&uuid)
+    }
+
+    pub fn devices_paths(&self) -> Vec<String> {
+        self.devices.values().map(|p| p.to_string()).collect()
```

Comment:
to be honest this looks quite inconsistent for me. connection path return `&str` and connection_paths `String`. I think it should ideally unify on one.

Reply:
Actually, both of them should return `OwnedObjectPath` objects. I will adapt it.